### PR TITLE
Remove `onOverrideMethodInvocationVarargsArrayNullability` handler method

### LINKED ISSUE #1501: Remove new `onOverrideMethodInvocationVarargsArrayNullability` handler method
Follow-up to #1485.  A better way to do this is to just use `Handler#onOverrideMethodType` which should already update the varargs array type with the appropriate nested nullability, along with other extant handler methods.

### PR DESCRIPTION
**Pull request by msridhar**
Fixes #1501 

Instead, we return varargs array nullability through the previous `onOverrideMethodInvocationParametersNullability` handler method.  We introduce a new `MethodParameterNullness` type to encapsulate regular parameter nullability and nullability of the varargs array; `MethodParameterNullness` is mutable for efficiency (like the arrays we used before).

This PR should not change visible behavior.  Still, I wasn't sure about some corner cases involving lambdas and method overriding, so there are some new tests.  I confirmed those new tests passed even without these changes.

## Summary by CodeRabbit

* **Refactor**
  * Reworked how method-parameter nullness is represented and passed through the analysis pipeline, simplifying per-parameter and varargs nullness handling.
  * Consolidated varargs-array nullness into the parameter nullness model and removed a separate varargs hook.

* **Tests**
  * Added JSpecify-mode tests exercising varargs and lambda parameter nullness inference.

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/handlers/RestrictiveAnnotationHandler.java [hunk 1]
```diff
@@ -104,50 +105,41 @@ private CodeAnnotationInfo getCodeAnnotationInfo(Context context) {
...
-      boolean isAnnotated,
-      @Nullable Nullness varargsArrayNullness) {
-    if (methodSymbol.isVarArgs()) {
-      Symbol.VarSymbol varargsParamSymbol =
-          methodSymbol.getParameters().get(methodSymbol.getParameters().size() - 1);
-      if (Nullness.varargsArrayIsNonNull(varargsParamSymbol, config)) {
-        return Nullness.NONNULL;
+      if (isVarargsParam && Nullness.varargsArrayIsNonNull(varargsParamSymbol, config)) {
+        argumentNullness.setVarargsArrayNullness(Nullness.NONNULL);
       }
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/RestrictiveAnnotationHandler.java:141]: The restrictive annotation handler only handles `@NonNull` annotations in null-unmarked (unannotated) code.  So I think this is fine.

File: nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyVarargsTests.java [hunk 2]
```diff
@@ -694,6 +694,88 @@ public void varargs(@Nullable Object @Nullable... params) {
...
+            import org.jspecify.annotations.Nullable;
+            @NullMarked
+            class Test {
+              @FunctionalInterface
+              interface NullableVarargsElementsConsumer {
+                void apply(@Nullable Object... inputs);
+              }
+              void test() {
+                NullableVarargsElementsConsumer consumer =
+                    inputs -> {
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyVarargsTests.java:741]: So, the semantics here is that `inputs` is the full array? Can't be called as `i1, i2, i3 -> { ... }`, then? (I assume so, just double checking)

File: nullaway/src/main/java/com/uber/nullaway/MethodParameterNullness.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package com.uber.nullaway;
+
+import com.sun.tools.javac.code.Symbol;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Mutable nullness information for a method's formal parameters and varargs array. For a varargs
+ * method, the nullability of individual arguments is represented in the last parameter position,
+ * while the nullability of the varargs array itself is represented separately.
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/MethodParameterNullness.java:9]: Do we plan to allow using this class in the future to override deep nullability? E.g. allow stuff like overriding argument 1 of `foo(List<Bar> bars)` to be `List<@Nullable Bar>` or `List<? extends @Nullable Bar>` or whatever? Obviously outside of the scope of this PR, but I am curious...

We currently have a separate model from LibraryModels to set these kinds of annotations for third-party libraries / JDK method, right? (using astub(x)?)

**Review by lazaroclapp — APPROVED** [review]
LGTM!

Comments below are just random related questions

File: nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyVarargsTests.java (hunk 2 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyVarargsTests.java:741]: Actually, it looks like for a varargs functional-interface method, you can only assign a lambda that takes a single argument, which must be an array; this example doesn't compile https://godbolt.org/z/YsYP1G6KT

File: nullaway/src/main/java/com/uber/nullaway/MethodParameterNullness.java (hunk 3 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/MethodParameterNullness.java:9]: Right now this is handled by the `onOverrideMethodType` handler method, see, e.g., here for library models:

https://github.com/uber/NullAway/blob/7b5c9243f89a7a344025006942a0378500d2e954/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java#L453

It would be great to unify the two eventually, but it will also be painful given that we still need to handle declaration annotations and other legacy corner cases.  Plus, all the original APIs didn't handle nested annotations, and that support has been tacked on for JSpecify mode, in a way that allows for JSpecify mode to be turned off.  I think maybe when JSpecify mode is the default, or even better when there is only JSpecify mode for NullAway, we could think about cleaning things up into one handler method.
